### ISSUE TITLE
URL with percent-encoded characters breaks router

### ISSUE THREAD

**Issue by cmnjcs** [rackt/react-router#1099]
Using v0.13.2. We have the following route declarations:

    <Route name="RootRoute" handler={App} path="/">
      <Route name="SSORoute" handler={SSO} path="sso" />
      [removed for conciseness]
    </Route>

This matches `http://hostname/sso?sso=bm9uY2U9NzJkOWZmM2NjMWNkMzM1OTliYWY4MTJiNGQyOTczN2QmcmV0dXJu`. However, when I attempt to navigate to `http://hostname/sso?sso=bm9uY2U9NzJkOWZmM2NjMWNkMzM1OTliYWY4MTJiNGQyOTczN2QmcmV0dXJu%0AX3Nzb191cmw9aHR0cCUzQSUyRiUyRmNvbW11bml0eS5ncm92ZWxhYnMuaW8l%0AMkZzZXNzaW9uJTJGc3NvX2xvZ2lu%0A&sig=5f8a827b0b4e6d56aa3473db01f45fb35ca165b742722bda7a043d477e5951d6` (the full URL, which includes percent-encoded characters) I get a white page. No errors reported on the client and server. I can't find any page that's being rendered as a result of going to that URL. Is this possibly related to https://github.com/rackt/react-router/issues/650?

**Comment by whatupdave** [comment]
I'm getting this too.

**Comment by whatupdave** [comment]
It's the linefeed character %0A. Sidestepping the issue at the moment but stripping them out of the url before redirecting here.

**Comment by cmnjcs** [comment]
Did some experimenting in our app. It looks like the issue could be fixed in this method, in [Match](https://github.com/rackt/react-router/blob/master/modules/Match.js#L48-L57), by adding the call to `replace()`:

    static findMatch(routes, path) {
      path = path.replace(/\n/g,"");
      var pathname = PathUtils.withoutQuery(path);
      var query = PathUtils.extractQuery(path);
      var match = null;

      for (var i = 0, len = routes.length; match == null && i < len; ++i)
        match = deepSearch(routes[i], pathname, query);

      return match;
    }

Thoughts?

**Comment by cmnjcs** [comment]
Whoops, that doesn't work, because it appears that it results in the queries being modified.

**Comment by cmnjcs** [comment]
Ok, different idea. Replace line breaks with the `%0A` characters, as shown below. I feel like this is not the right place to do this, or the right way to be doing this, but I will say that it works for my application (as far as I can tell).

    value: function findMatch(routes, path) {
      path = path.replace(/\n/g,"%0A");
      var pathname = PathUtils.withoutQuery(path);
      var query = PathUtils.extractQuery(path);
      var match = null;

      for (var i = 0, len = routes.length; match == null && i < len; ++i) match = deepSearch(routes[i], pathname, query);

      return match;
    }

**Comment by aindlq** [comment]
I think I've traced the issue. The problem is with PathUtils.queryMatcher:
```
var queryMatcher = /\?(.*)$/;
```
. doesn't match newlines in regex, so it should be:
```
var queryMatcher = /\?(.|\n)*/;
```

**Comment by cmnjcs** [comment]
Any updates on this?

**Comment by ryanflorence** [comment]
would love to verify if this is still an issue on master (1.0 API has changed a bit), can you verify, and if its still a problem, a failing test would be awesome

**Comment by cmnjcs** [comment]
Trying to investigate, but not sure how long the refactor to get rid of `Router.run()` will take.

**Comment by mjackson** [comment]
@cmnjcs We're wrapping up the rewrite, getting ready to ship 1.0 very soon. The code on master is fairly stable right now. I don't anticipate any major changes between here and 1.0.

**Comment by cmnjcs** [comment]
Sorry for the lack of clarity--when I said refactor, I meant of our application. The examples in the [docs](https://rackt.github.io/react-router/tags/v1.0.0-beta2.html) and the [README](https://github.com/rackt/react-router) (as a heads up, the examples are different) don't work in our application, so trying to restore functionality so I can test if the bug has been fixed.

**Comment by mjackson** [comment]
This was fixed in #1166

**Comment by cmnjcs** [comment]
Thanks! :+1:
